### PR TITLE
JIT: rework optCanonicalizeLoop

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Rework loop canonicalization in anticipation of extending it to handle
non-loop edges like those seen in #70569.

The code to fix up flow out of `h` was not needed and has been removed;
we now assert upstream that no such fix up will be needed.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

No diffs expected.

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -2874,38 +2874,49 @@ bool Compiler::optIsLoopEntry(BasicBlock* block) const
...
-
-    // Is the top of the current loop in any nested loop?
-    if (optLoopTable[loopInd].lpTop->bbNatLoopNum != loopInd)
-    {
-        if (optCanonicalizeLoop(loopInd))
-        {
-            modified = true;
-        }
-    }
+    bool modified = optCanonicalizeLoop(loopInd);
```
Comment (kunalspathak) [inline src/coreclr/jit/optimizer.cpp:2894]: I didn't understand why the check of `if (optLoopTable[loopInd].lpTop->bbNatLoopNum != loopInd)` was removed although the comment says `"For loopInd and all contained loops, ensures each loop top's back edges do not come from nested loops."`

Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:2894]: We were checking this redundantly; now we just check it down in `optCanonicalizeLoop`.

I may move that check back up to the caller with my next round of revisions, but the point is we don't need to check it twice.

Comment (kunalspathak) [reply src/coreclr/jit/optimizer.cpp:2894]: ah..see it now.

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/coreclr/jit/optimizer.cpp [hunk 2]
```diff
@@ -2874,38 +2874,49 @@ bool Compiler::optIsLoopEntry(BasicBlock* block) const
...
 
     for (unsigned char child = optLoopTable[loopInd].lpChild; //
          child != BasicBlock::NOT_IN_LOOP;                    //
          child = optLoopTable[child].lpSibling)
     {
-        if (optCanonicalizeLoopNest(child))
-        {
-            modified = true;
-        }
+        modified |= optCanonicalizeLoopNest(child);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:2900]: fwiw, I prefer the previous expression.

File: src/coreclr/jit/optimizer.cpp [hunk 3]
```diff
@@ -3016,9 +3043,11 @@ bool Compiler::optCanonicalizeLoop(unsigned char loopInd)
     // a call to fgUpdateChangedFlowGraph which will recompute the reachability sets anyway.
 
     // Redirect the "bottom" of the current loop to "newT".
-    BlockToBlockMap* blockMap = new (getAllocator(CMK_LoopOpt)) BlockToBlockMap(getAllocator(CMK_LoopOpt));
+    BlockToBlockMap* const blockMap = new (getAllocator(CMK_LoopOpt)) BlockToBlockMap(getAllocator(CMK_LoopOpt));
     blockMap->Set(t, newT);
     optRedirectBlock(b, blockMap);
+    JITDUMP("Redirecting top->bottom backedge " FMT_BB " -> " FMT_BB " to " FMT_BB " -> " FMT_BB "\n", b->bbNum,
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:3049]:
```suggestion
    JITDUMP("Redirecting bottom->top backedge " FMT_BB " -> " FMT_BB " to " FMT_BB " -> " FMT_BB "\n", b->bbNum,
```

Comment (BruceForstall) [reply src/coreclr/jit/optimizer.cpp:3049]: Maybe we should just add this output to `optRedirectBlock` at the point where `bbJumpDest` or `>bbJumpSwt->bbsDstTab[i]` is set.

File: src/coreclr/jit/optimizer.cpp [hunk 4]
```diff
@@ -3039,9 +3068,14 @@ bool Compiler::optCanonicalizeLoop(unsigned char loopInd)
         // outside-in, so we shouldn't encounter the new blocks at the loop boundaries, or in the predecessor lists.
         if (t->bbNum <= topPredBlock->bbNum && topPredBlock->bbNum <= b->bbNum)
         {
-            JITDUMP("in optCanonicalizeLoop: 'top' predecessor " FMT_BB " is in the range of " FMT_LP " (" FMT_BB
-                    ".." FMT_BB "); not redirecting its bottom edge\n",
-                    topPredBlock->bbNum, loopInd, t->bbNum, b->bbNum);
+            // Note we've already redirected t->b above.
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:3071]:
```suggestion
            // Note we've already redirected b->t above.
```

**Review by BruceForstall — CHANGES_REQUESTED** [review]
Are there diffs?

File: src/coreclr/jit/optimizer.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:2900]: Odd that reply ended up there.

In the follow-on changes there will be more reasons to canonicalize loops and pre-screening for them all starts to get clunky.

File: src/coreclr/jit/optimizer.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:3049]: We could, but we won't have a friendly name for the edge unless the caller passes one.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**